text/0087-graphql-errors.md

```diff
@@ -0,0 +1,260 @@
+
+```json
+{
+  "contexts": {
+    "response": {
+      "type": "response",
+      "status_code": 500,
+      "body_size": 1000,
+      "data": "...",
+    }
+  }
+}
+```
+
+* `data`: Can be given as string or structural data of any format..
+
+The `Response` interface keeps arbitrary fields, it's back compatible with the current implementation.
+
+## Must have
+
+The fields `Request#data` and `Response#data` could contain PII and they should run data scrubbing agressively.
+
+[Session Replay](https://docs.sentry.io/platforms/javascript/guides/remix/session-replay/configuration/) already sends the request and response bodies, so we can use the same data scrubbing rules.
+
+Since GraphQL is a well defined spec, we can also scrub the GraphQL fields.
+
+Request example:
+
+```json
+{
+  "query": "{\n  viewer {\n    login\n  }\n}",
+  "variables": {
+    "login": "marandaneto"
+  }
+}
+```
+
+Response example:
+
+```json
+{
+  "data": {
+    "viewer": {
+      "login": "[Filtered]"
+    }
+  }
+}
+```
+
+In this case, we only need to use the Request `variables` and its keys to scrub the Response `data`.
+
+# Drawbacks
+
+[Envelopes](https://develop.sentry.dev/sdk/envelopes) (Events) contain way lower [size limits](https://develop.sentry.dev/sdk/envelopes/#size-limits). The `data` fields could be large and it could be a problem.
+
+SDKs should discard large and binary bodies by default, using the [maxRequestBodySize](https://docs.sentry.io/platforms/android/configuration/options/#max-request-body-size) and `maxResponseBodySize` (it'll be added) options.
+
+The difference is that for GraphQL errors, this should be enabled by default.
+
+# Appendix
+
+## Removed Proposals
+
+### Option 2
+
+Add a new `graphql` interface to [Contexts](https://develop.sentry.dev/sdk/event-payloads/types/).
+
+```json
+{
+  "contexts": {
+    "graphql": {
+      "type": "graphql",
+      "data": "...",
+    }
+  }
+}
+```
+
+We'd need to duplicate or still use some fields from the `Request` and `Response` interface.
+
+Size limits would still be a problem.
+
+### Option 3
+
+Add a new envelope item for GraphQL.
+
+```
+{"event_id":"9ec79c33ec9942ab8353589fcb2e04dc"\n
+{"type":"graphql","length":41,"content_type":"application/json"}\n
+{"request":"foo","response":"bar"}
+```
+
+This would not be back compatible and must be added to all SDKs.
+
+### Option 4
+
+Add Request and Response bodies as attachments.
+
+```
+{"event_id":"9ec79c33ec9942ab8353589fcb2e04dc"\n
+{"type":"attachment","length":10,"content_type":"application/json","filename":"request.txt"}\n
+foo
+{"type":"attachment","length":10,"content_type":"application/json","filename":"response.txt"}\n
+bar
+```
+
+Attachments have to be special cased in Sentry, seems hacky, we do that with screenshots already.
+
+# Unresolved questions
+
+- Do we need to send the GraphQL scheme to Sentry in order to do data scrubbing properly?
+- Should this feature be opt-in due to the size of the data and PII risk?
```

Comment:
I would strongly advise making this an opt-in feature only.

Reply:
Added a note.